ingestion_server/ingestion_server/cleanup.py

```diff
@@ -266,30 +267,96 @@ def _clean_data_worker(rows, temp_table, sources_config, all_fields: list[str]):
+    part: int
+    rows: list[tuple[str, str]]
 
-    cleanup_counts = {field: len(items) for field, items in result.items()}
-    for field, cleaned_items in result.items():
-        # Skip the tag field because the file is too large and fills up the disk
-        if field == "tag":
-            continue
-        if cleaned_items:
-            with open(f"{field}.tsv", "a") as f:
-                csv_writer = csv.writer(f, delimiter="\t")
-                csv_writer.writerows(cleaned_items)
 
-    end_time = time.time()
-    total_time = end_time - start_time
-    log.info(f"Finished saving cleaned data in {total_time},\n{cleanup_counts}")
-    return cleanup_counts
+class CleanDataUploader:
+    # Number of records to buffer in memory before writing to disk
+    mem_buffer_size = 100
+
+    # Size (in MB) of local file buffer before writing to S3
+    disk_buffer_size = 850 * 1024 * 1024
+
+    s3_path = "shared/data-refresh-cleaned-data"
+
+    buffer = {
+        field: FieldBuffered(part=1, rows=[])
+        for field in _cleanup_config["tables"]["image"]["sources"]["*"]["fields"]
+    }
+
+    def __init__(self):
+        bucket_name = config("AWS_S3_BUCKET", default="openverse-catalog")
+        self.s3 = self._get_s3_resource()
+        self.s3_bucket = self.s3.Bucket(bucket_name)
+        self.date = time.strftime("%Y-%m-%d")
+
+    @staticmethod
+    def _get_s3_resource():
+        if config("ENVIRONMENT", default="local") == "local":
+            return boto3.resource(
+                "s3",
+                endpoint_url=config("AWS_S3_ENDPOINT", default="http://s3:5000"),
+                aws_access_key_id=config("AWS_ACCESS_KEY_ID", default="test_key"),
+                aws_secret_access_key=config(
+                    "AWS_SECRET_ACCESS_KEY", default="test_secret"
+                ),
+            )
+
+        return boto3.resource(
+            "s3", region_name=config("AWS_REGION", default="us-east-1")
+        )
+
+    def _save_to_disk(self, field: str, force_upload=False):
+        log.info(
+            f"Saving {len(self.buffer[field].rows)} rows of `{field}` to local file."
+        )
+        filename = f"{field}.tsv"
+
+        with open(filename, "a") as f:
+            csv_writer = csv.writer(f, delimiter="\t")
+            csv_writer.writerows(self.buffer[field].rows)
+            # Clean memory buffer of saved rows
+            self.buffer[field].rows = []
+
+        if os.path.getsize(filename) >= self.disk_buffer_size or force_upload:
+            try:
+                self._upload_to_s3(field)
+            except Exception as e:
+                log.error(f"Error uploading {field} to S3: {e}")
+
+    def _upload_to_s3(self, field: str):
+        part_number = self.buffer[field].part
+        log.info(f"Uploading file part {part_number} of `{field}` to S3...")
+        s3_file_name = f"{self.s3_path}/{self.date}_{field}_{part_number}.tsv"
+        self.s3_bucket.upload_file(f"{field}.tsv", s3_file_name)
```

Comment:
Maybe extract the tsv file name here, too?
```suggestion
        tsv_file_name = f"{field}.tsv"
        self.s3_bucket.upload_file(tsv_file_name, s3_file_name)
```

Reply:
Done 👌